Source/Core/VideoBackends/Software/Rasterizer.cpp

```diff
@@ -265,11 +301,46 @@ static void BuildBlock(s32 blockX, s32 blockY)
   }
 }
 
+void PrepareTriangleFrontFace(const OutputVertexData* v0, const OutputVertexData* v1,
+                              const OutputVertexData* v2)
+{
+  // This is mainly used for updating the zslope, as zfreeze requires it to be updated even if the
+  // triangle is culled.
+  const s32 X1 = iround(16.0f * v0->screenPosition[0]) - 9;
+  const s32 Y1 = iround(16.0f * v0->screenPosition[1]) - 9;
+
+  const SlopeContext ctx(v0, v1, v2, (X1 + 0xF) >> 4, (Y1 + 0xF) >> 4);
+
+  float w[3] = {1.0f / v0->projectedPosition.w, 1.0f / v1->projectedPosition.w,
+                1.0f / v2->projectedPosition.w};
+  WSlope = Slope(w[0], w[1], w[2], ctx);
+
+  if (!bpmem.genMode.zfreeze)
+    ZSlope = Slope(v0->screenPosition.z, v1->screenPosition.z, v2->screenPosition.z, ctx);
+
+  for (unsigned int i = 0; i < bpmem.genMode.numcolchans; i++)
+  {
+    for (int comp = 0; comp < 4; comp++)
+      ColorSlopes[i][comp] = Slope(v0->color[i][comp], v1->color[i][comp], v2->color[i][comp], ctx);
+  }
+
+  for (unsigned int i = 0; i < bpmem.genMode.numtexgens; i++)
+  {
+    for (int comp = 0; comp < 3; comp++)
+    {
+      TexSlopes[i][comp] = Slope(v0->texCoords[i][comp] * w[0], v1->texCoords[i][comp] * w[1],
+                                 v2->texCoords[i][comp] * w[2], ctx);
+    }
+  }
```

Comment:
For performance reasons, it might be worth splitting this function in half, one for depth, one for color/uv slopes and then only calculating the depth slope for culled triangles. 

In #10260 I got a speedup for delaying the calculation of color and uv values for each pixel until the depth passed for the quad. I wonder if it might be worth delaying the calculation of these slopes until the first pixel quad passes depth testing.

Reply:
I've renamed the function to `UpdateZSlope` and only update the zslope in it now (while the other slopes happen on their own); this does mean that `SlopeContext` is created twice though (one in `UpdateZSlope` and one in `DrawTriangleFrontFace`).

I feel like trying to delay the calculation of the slopes would cause more problems than it would fix, as unlike computing the slope values for each pixel, you only need to calculate the slopes once, so you'd need to check whether the slopes had already been computed for each block, which seems like it'd almost always be slower than computing the slopes once.

Reply:
It's something you would only do if profiling proved that it was worth it.

It's pretty cheap to check if the slopes have been setup for this triangle yet, a single load, a compare and a branch. What's potentially expensive is the extra branch misspredicticts. But branch prediction is really good on modern CPUs, taking branch history into account.... It's really hard to tell without profiling.

Actually, I've just realized such an optimisation will be made irrelevant by my plans to put a multithreadding boundary here, running the depth testing on the main thread to get timing information, then sending the depth test mask along with the triangles to multiple worker threads where the color will be calculated.